feature/token

Eomm opened the pull request:
**Type:**

First draft to implement issue #541 

**Description:**

This is a draft PR in order to implement the `npm token` feature.
I mocked the endpoints with the right input/output and tried with npm-cli 6.4.1

The commands are:
```
npm token list
   => return the list of the token created by the user
npm token revoke <tokenKey>
   => delete a token if the user has publish right
npm token create [--read-only] [--cidr=list]
   => create a token for publish and/or read only the packages
```

Impact on existing codebase:

* Auth: a bearer token must be validated as the actual user and password
* Storage: are necesary 3 new functions:
  * saveToken(IToken)
  * getTokens(username: string) (not sure if paginate)
  * deleteToken(tokenKey: string)

Talking with @juanpicado the process for closing this PR should be:

* [ ] edit the types of `@verdaccio/type`
* [ ] edit the `local-storage`
* [ ] update this repo with new types and complete the implementation of the APIs + tests

I may ask:

* I'm not skilled in typescript so could you give me an help in the first task?
* for the second task of storage the suggestion is to explore [leveldb](https://github.com/google/leveldb) and don't change the existing datastore. I don't analyze deeply this task yet. I check that `npm` doesn't save the token, it store only the first 4 char, and last 4 and then it generates a <tokenKey> to delete it if necessary.
* there are some guidelines to setup this project because `npm test` (after an `npm run build:webui`) give me this on master branch:

```
Test Suites: 10 failed, 1 skipped, 33 passed, 43 of 44 total
Tests:       15 failed, 5 skipped, 366 passed, 386 total
Snapshots:   9 failed, 17 passed, 26 total
Time:        70.205s
```

juanpicado commented:
> I'm not skilled in typescript so could you give me an help in the first task?

Count with me. 

>  I check that npm doesn't save the token, it store only the first 4 char, and last 4 and then it generates a to delete it if necessary.

I think that's the case, they only display it in the UI the token, and store exactly what you say (I suppose), but keep on mind we have 2 kind of token signatures in V4 https://github.com/verdaccio/verdaccio/issues/168#issuecomment-394149624 
So for this topic, I think I should give you some insigths how it works. I guess chat over Discord might be faster.

> there are some guidelines to setup this project because npm test (after an npm run build:webui) 

I'm about to remove the UI from this project https://github.com/verdaccio/verdaccio/pull/1252

juanpicado commented:
@Eomm I'd recommend fetching from upstream branch

Eomm commented:
The types that I think would be necessary to define are:

```
Token {
  user: 'john doo' // string: the creator of the token
  viewToken: '1234......1234' // string: the maschered string token
  key: 'autogenerated' // string: a generated hash from the token
  cidr: ['192.168.0.1', '10.10.10.10'] // Array<string> IP allow list for cidr_whitelist parameter
  readonly: true // boolean
  created: '2018-10-05T21:49:39.457Z' // Date - stored in UTC
  updated?: '2018-10-05T21:49:39.457Z' // Date - stored in UTC
}

TokenFilter {
  user: 'foo' // string: the user that want retrieve his token
}
```

For the key generation I would use this pattern to start:
```
token = UUIDv4() // as npm
parts = token.split('-')
// npm key is 128 char
key = parts.slice(1, 4).reduce((k, tokenPiece) => k += MD5(tokenPiece), parts[0])
```

The new storage functions could be:

```
saveToken(token: Token)
deleteToken(user: string, tokenKey: string)
getTokens(filter: TokenFilter)
```

Do you agree?

juanpicado commented:
> For the key generation I would use this pattern to start:

We need to discuss first the token.

The token at npmjs looks like this format
````
xxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx
````
while our token at `v3` looks like
````
"IRdsdsdSvZE/mdsdsdsdsA=="
```
and using the new JWT (https://jwt.io/)
````
eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiaWF0IjoxNTE2MjM5MDIyfQ.SflKxwRJSMeKKF2QT4fwpMeJf36POk6yJV_adQssw5c
````

In the case of Verdaccio 4, we are dealing with 2 sorts of formats (I'm writing a blog post about this that I'll share here so you have a better overview of how it works)  You can see the PR here https://github.com/verdaccio/verdaccio/pull/896

I'd recommend check the method `apiJWTmiddleware()` in `auth.js` 

```
 if (isAESLegacy(security)) {
        this.logger.trace('api middleware using legacy auth token');
        this._handleAESMiddleware(req, security, secret, authorization, next);
      } else {
        this.logger.trace('api middleware using JWT auth token');
        this._handleJWTAPIMiddleware(req, security, secret, authorization, next);
      }
```

Based on the token signature selected by the user we must sign the token in the same way, if we do not follow the same process, tokens will be invalid on going through the middleware.

juanpicado commented:
Perhaps you can give me more context about 

```
getTokens(filter: TokenFilter)
```
Because I thought, we were storing only small part o the token, but we do not store the full token, right?

juanpicado commented:
Aside of that, the API loos good to me

```
saveToken(token: Token)
deleteToken(user: string, tokenKey: string)
getTokens(filter: TokenFilter) // just waiting more context
```

Perhaps we do not use `cidr`, I dunno, what do you think?.  But the structure seems good to me
```
Token {
  user: 'john doo' // string: the creator of the token
  viewToken: '1234......1234' // string: the maschered string token
  key: 'autogenerated' // string: a generated hash from the token
  cidr: ['192.168.0.1', '10.10.10.10'] // Array<string> IP allow list for cidr_whitelist parameter
  readonly: true // boolean
  created: '2018-10-05T21:49:39.457Z' // Date - stored in UTC
  updated?: '2018-10-05T21:49:39.457Z' // Date - stored in UTC
}

TokenFilter {
  user: 'foo' // string: the user that want retrieve his token
}
```

Eomm commented:
> In the case of Verdaccio 4, we are dealing with 2 sorts of formats
> tokens will be invalid on going through the middleware.

Oh, sorry, I was wondering that we should add the UUIDv4 because those 2 auth token could be decrypted and UUID is a randomly generated token.
So the question will be: what we will write in those tokens?

> Perhaps you can give me more context about

`npm` for every token actions (list, create, revoke) ask for the list of the token created by the user (`npm token list`):
+ list => to show the list of the tokens created by the user
+ create => I don't know why 😅
+ revoke => before calling the `HTTP DELETE` npm will check if the token-key you want to delete is in your tokens list
So that function is needed to get all the (masquerade) tokens of the user

> Because I was thinking we were storing only small part o the token, but we do not store the full token, right?

Sure, we must not store all the tokens.
My original idea was:
on `npm token create`
+ create a random token always valid => save the HASH + the maskered token
+ return the random token to the npm client so the user can see it

on `npm install/publish my-module`
+ a request comes with a token, I generate the HASH and check if it exists in my store*
+ if it exists I check if the user who generated it has the read/write grant to the package requested
+ if the user has the right, I authorize the request to download/publish the package
----
but if we use JWT or the other one the logic should be more simple:

on `npm token create`
+ create a token always valid => save only the masquerade token
+ return the token to the npm client so the user can see it and store it

on `npm install/publish my-module`
+ a request comes with a token, the auth middleware will validate it

So your solution is more straight 👍

Eomm commented:
> Perhaps we do not use cidr, I dunno, what do you think?

This feature should be added to the auth middleware if we use the JWT I think.

I'll try think loudly:
+ if the token is not a login, but a "generated token" we should query the data-store to get the "token details".
+ if the token details have cidr we must validate also the request IP addrs

This is a security task, I can't evaluate how much cost add it, but for sure it could be added in future, so for this reason I would store those values also we don't use them at the beginning

juanpicado commented:
hi @Eomm , let's beings with `npm create` . I drew this for you. It is more understandable the whole flow?

![npm_token_create 001](https://user-images.githubusercontent.com/558752/56091048-08681180-5eaa-11e9-9e59-5d9cea9c4806.jpeg)

As you can see, we do not have to create the token, verdaccio has utilities for it out of the box, you only need to use them. Let me know whether you need more context.

Let's do `create` first and then we go with other commands.

Eomm commented:
Great schema! I will start write it down during the week 👍🏼

Thank you very much for the help 💪🏻

juanpicado commented on src/api/endpoint/api/v1/token.js:
```diff
@@ -0,0 +1,112 @@
+    const { password, readonly, cidr_whitelist } = req.body;
+    const { name } = req.remote_user;
+
+    if (typeof readonly !== 'boolean' || !Array.isArray(cidr_whitelist)) {
```
Here for consistency with the rest of the code, please use `lodash`

juanpicado replied:
The same in other `typeof`

juanpicado commented on src/api/endpoint/api/v1/token.js:
```diff
@@ -0,0 +1,112 @@
+
+          const saveToken = {
+            user: name,
+            viewToken: `${token.substr(0, 3)}...${token.substr(-3)}`,
```
`${token.substr(0, 3)}...${token.substr(-3)}` this should be in a utility thus we can test it easily. We had to add unit test at the end of all of this.

juanpicado commented:
@Eomm all ok? Any support need it here?

Eomm commented:
Hi, I will go further soon: I checked out local-storage to implement the new functions

juanpicado commented:
Perfect 👏 , do not hesitate to reach me out if you need more help.

juanpicado commented on src/api/endpoint/api/v1/token.js:
```diff
@@ -0,0 +1,113 @@
+import logger from '../../../../lib/logger';
+
+import type { $Response, Router } from 'express';
+import type { $NextFunctionVer, $RequestExtend, IAuth } from '../../../../../types';
```
Some types are missing here. 
```
import type { Config } from '@verdaccio/types';
import type { $NextFunctionVer, $RequestExtend, IAuth, IStorageHandler } from '../../../../../types';
```

juanpicado commented on src/api/endpoint/api/v1/token.js:
```diff
@@ -0,0 +1,113 @@
+          const token = await getApiToken(auth, config, user, password);
+          const key = stringToMD5(token);
+
+          const saveToken = {
```
`const saveToken: Token = {`

Eomm commented:
I will fix all the comments as final rush when I complete the tasks in local-storage and types 👍

juanpicado commented:
New types published "@verdaccio/types": "6.0.2"

juanpicado commented:
@Eomm don't forget to update the branch on a regular basis.

juangabreil commented:
Hi, is there a way I can help you with this feature? I'm very interested on adding support for this to enable my CI/CD pipeline.

juanpicado commented:
hi @juangabreil , first let's ask @Eomm , he is on a charge of this.  Do you need some help?

Eomm commented:
Hi @juangabreil 
right now I'm unable to find the right amount of time to solve this PR:
https://github.com/verdaccio/local-storage/pull/165

I'll try to explain:

1. I defined the interface of the new methods to store and get the token (ref https://github.com/verdaccio/flow-types/pull/62 )
1. Now we should implement that interface. I started [here](https://github.com/verdaccio/local-storage/pull/165) but I don't have the right amount of time to search where store the tokens in a secure way. (I need to say that we must not store the token, but only and hash)
1. Finally, complete this PR using the implementation of the `local-storage`

I can help with tasks that are less "time spending", otherwise I will have the time for this analysis after this month 😔

juangabreil commented:
Ok, I will start working on that :)

juanpicado commented:
I close this one, follow up https://github.com/verdaccio/verdaccio/pull/1427
